### ISSUE TITLE
Custom GORM graphQL dataFetcher's lack a Session beyond the dataFetcher's closure

### ISSUE THREAD

**Issue by jamesdh** [micronaut-projects/micronaut-groovy#199]
**TLDR:** 
So it appears this wasn't an issue in Grails/Spring because of their default use of an `OpenSessionInViewInterceptor`. Yes, `DefaultGormDataFetcher` wraps it's queries in a transaction. This issue appears to be due to the combination of some sort of session/transaction demarcation that occurs when you define your own dataFetcher (both `ClosureDataFetcher` or anything that extends `DefaultGormDataFetcher`, etc) and Micronaut's lack of its own `OpenSessionInViewInterceptor`

Just to be clear, I'm not suggesting Micronaut implement an equivalent to OpenSessionInViewInterceptor, I'm just pointing out that *something* has to be done in order for all queries to work correctly. 

**Longer Version:** 
Let's assume we have a `Library` -> `Author` -> `Book` object model defined as 1-to-many between each. Let's assume there are no custom datafetchers defined within this example.

When invoking a deep graph query, for example:

```GraphQL
query {
  library(id: "xyz") {
    name
    address
    authors {
      dateOfBirth
      books {
        isbn
        title
      }
    }
  }
}
```

This works as expected. However, when using a custom dataFetcher, for example on the Library model: 

```groovy
static graphql = GraphQLMapping.build {
    operations.all.enabled false
    property('authors') {
        dataFetcher { Library source, ClosureDataFetchingEnvironment env ->
            Author.where { 
              //some criteria...
            }
        }
    }
}
```

Then the initial Library is correctly retrieved, but upon retrieving the Author we get:

    org.hibernate.HibernateException: No Session found for current thread

You can change it up so that instead of returning a `DetachedCriteria`, you instead return retrieved objects but the same issue arrises upon GraphQL's attempt to fetch the remaining associations. 

Likewise, when defining a *regular* DataFetcher, for example:

```groovy
query('library', Library) {
    dataFetcher(new SingleEntityDataFetcher<Library>(Library.gormPersistentEntity) {
        Library get(DataFetchingEnvironment env) {
            (Library)withTransaction(true) {
                // We previously stashed the authenticated User on the GraphQLContext...
                GraphQLContext context = (GraphQLContext)env.context
                User user = (User)context.get('requester')
                user.defaultLibrary
            }
        }
    })
}
```
and trying to do a similar query as above, we then get...

    failed to lazily initialize a collection of role: Library.authors, could not initialize proxy - no Session

However, if instead of overriding the `get()` method we instead override the `buildCriteria()` method on our DataFetcher and return a DetachedCriteria, a la:

```groovy
query('library', Library) {
    dataFetcher(new SingleEntityDataFetcher<Library>(Library.gormPersistentEntity) {
        protected DetachedCriteria buildCriteria(DataFetchingEnvironment env) {
            // We previously stashed the authenticated User on the GraphQLContext...
            GraphQLContext context = (GraphQLContext)env.context
            User user = (User)context.get('requester')
            Library.where { id == user.defaultLibraryId }
        }
    })
}
```

Then it correctly traverses the entire graph query and is able to fetch all properties/associations. 

So because of how transactional/session boundaries apparently differ between regular `DataFetcher`'s and `ClosureDataFetcher`'s, and because Micronaut no longer has an OpenSessionInViewInterceptor, many of the examples as shown in the GORM-graphql docs simply do not work. 

Since I'm also using `micronaut-graphql` to supply the controller for receiving queries, I was able to work around the above by defining a GraphQLExecutionInputCustomizer which receives the HttpRequest and binds a Session to the thread, thus maintaining an open Session for the extent of the query. 

```groovy
@Singleton
class ContextBuilder implements GraphQLExecutionInputCustomizer {

    @Inject SessionFactory sessionFactory

    @Override
    Publisher<ExecutionInput> customize(ExecutionInput executionInput, HttpRequest httpRequest, @Nullable MutableHttpResponse<String> httpResponse) {

        Session session = sessionFactory.openSession()
        SessionHolder sessionHolder = new SessionHolder(session)
        TransactionSynchronizationManager.bindResource(sessionFactory, sessionHolder)
        //...
    }
}
```

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [ ] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

See above...

### Expected Behaviour

Examples as shown in GORM-graphql "just work"

### Actual Behaviour

Because Micronaut does not have any kind of OSIV behavior, many examples fall down, especially when utilizing ClosureDataFetcher's. 

### Environment Information

- **Operating System**: MacOS 10.15
- **Micronaut Version:** 2.3.3
- **JDK Version:** 11

### Example Application

Until an update of `micronaut-groovy` is released that contains some of the previous fixes in regards to gorm/gorm-graphql, making an example app isn't actually very feasible without jumping through some custom dependency hoops
